# 필수과제
- 오늘 시간까지 배운 내용을 타이타닉 데이터를 가지고 전처리 해주세요!
- mpg 데이터는 복습으로 진행
- 타이타닉데이터는 필수과제로 이용해 주세요!
- 단순한 과제가 아니라 타이타닉 데이터의 컬럼에  대한 도메인을 찾으시고 (구글링하면 바로 나옴 타이타닉 데이터 컬럼 설명)
- '생존'이 중요한 키워드라 y값
- 생존을 중심으로 다양한 컬럼들이 어떤 특성을 가지고 있는지?
- 본인이 질문을 최소 10개 이상 하고 그 질문에 답변을 할 수 있는 코드를 작성하고 주석처리까지 진행해 주세요.
- 그냥 코드만 복붙한 거는 과제로 인정하지 않습니다.
- 꼭 구글링은 최대한 지양하지만 필요하다면 진행해 주시고, 복습과제에서 추가로 하면서 실력을 키워보세요!
- 배운 코드로 응용하기

## 타이타닉 데이터 컬럼 설명
- **survived** 생존여부 (0:사망, 1:생존)
- **pclass** 객실 등급 (1:1등급, 2:2등급, ...)
- **sex** 성별 (male:남자, female:여자)
- **age** 나이
- **sibsp** 함께 탑승한 형제 및 배우자 수
- **parch** 함께 탑승한 자녀 및 부모 수
- **fare** 요금
- **embarked** 탑승지 이름의 앞글자
- **class** 객실 등급 (first:1등급, second:2등급, ...)
- **who** (man:남자, woman:여자, child:아이)
- **adult_male** 성인 남자 여부 (True:성인 남자, False:그외)
- **deck** 선실 번호 첫 알파벳
- **embark_town** 탑승지 이름
- **alive** 생존 여부 (yes:생존, no:사망)
- **alone** 혼자 탑승 여부 (True:혼자, False:가족과 함께)


In [9]:
import seaborn as sns

In [10]:
import pandas as pd

In [11]:
df = sns.load_dataset('titanic')

In [12]:
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

**1-1)  남자와 여자 중 어느쪽의 생존자 수가 더 많은가?**

**1-2)  남자와 여자 중 생존자의 비율이 더 큰쪽은 어디인가?**

In [14]:
df.query("survived == 1").groupby('sex').agg(survived = ('survived', 'count'))
#female(여자)와 male(남자)의 생존자 수

survived
sex             
female       233
male         109

In [30]:
df.query("survived == 1 and sex == 'male'").groupby('sex').agg(survival_rate = ('survived', 'count'))/df.query("sex == 'male'")['sex'].count()
#남자의 생존 비율 (생존한 남자의 수/전체 남자의 수)

survival_rate
sex                
male       0.188908

In [31]:
df.query("survived == 1 and sex == 'female'").groupby('sex').agg(survival_rate = ('survived', 'count'))/df.query("sex == 'female'")['sex'].count()
#여자의 생존 비율

survival_rate
sex                  
female       0.742038

In [35]:
df.groupby('sex')['survived'].mean() #쉬운 방법으로 표현한 것

sex
female    0.742038
male      0.188908
Name: survived, dtype: float64

생존자의 수는 남자보다 여자가 더 많다.
남자의 생존률은 0.188908이며 여자의 생존률은 0.742038이므로 여자의 생존 비율이 더 크다.

**2) 객실 등급에 따른 생존자 수와 비율은 어떠한가?**

In [32]:
df.query("survived == 1").groupby('pclass').agg(survived = ('survived', 'count'))
#pclass(객실 등급)에 따른 생존자의 수

survived
pclass          
1            136
2             87
3            119

In [34]:
df.groupby('pclass')['survived'].mean()
#객실 등급에 따른 생존자 비율

pclass
1    0.629630
2    0.472826
3    0.242363
Name: survived, dtype: float64

1등급 객실의 생존자 수가 136명으로 가장 많았으며 생존자 비율또한 0.629630으로 가장 컸다.

**3) 나이가 생존자 수와 연관이 있을까?**

In [43]:
df.query("survived == 1 and age <= 20")['survived'].count() 
#20세 이하 생존자의 수

82

In [44]:
df.query("survived == 1 and age > 20 and age <= 40")['survived'].count()
#20세 초과 40세 이하 생존자의 수

153

In [45]:
df.query("survived == 1 and age > 40 ")['survived'].count()
#40세 초과 생존자의 수

55

41세 이상인 사람의 생존자 수가 가장 적으며
21세 이상 40세 이하인 생존자 수가 가장 많다.
나이대를 구분한 범위가 넓지만 나이와 생존자의 수는 어느정도 연관이 있는 것으로 보인다.

**4) 남자아이와 여자아이의 생존자수와 비율은?**

In [47]:
df.query("survived == 1 and sex == 'male' and who == 'child'")['survived'].count()
#남자아이의 생존자 수

21

In [48]:
df.query("survived == 1 and sex == 'female' and who == 'child'")['survived'].count()
#여자아이의 생존자 수

28

In [53]:
df.query("survived == 1 and sex == 'male' and who == 'child'")['survived'].count()/df.query("who == 'child' and survived == 1")['who'].count()
#남자아이의 생존율 (생존한 남자아이 수/생존한 전체 아이 수)

0.42857142857142855

In [54]:
df.query("survived == 1 and sex == 'female' and who == 'child'")['survived'].count()/df.query("who == 'child' and survived == 1")['who'].count()
#여자아이의 생존율

0.5714285714285714

남자아이보다 여자아이의 생존자 수가 더 많았으며
이들의 생존율을 비교해봤을 때 여자아이의 생존율이 더 높았다.

**5) 남자 중 성인인 사람과 아닌 사람의 생존율은?**

In [64]:
df.query("survived == 1 and adult_male == True")['survived'].count()/df['adult_male'].count()
#성인 남성의 생존율 (생존한 성인 남성의 수/ 전체 성인 남성의 수)

0.09876543209876543

In [63]:
df.query("survived == 1 and adult_male == False")['survived'].count()/df.query("adult_male == False")['adult_male'].count()
#성인이 아닌 남성의 생존율

0.7175141242937854

남자 중 성인이 아닌 남성의 생존율이 더 높은 것을 확인할 수 있다.

**6) 선실 번호의 첫 알파벳에 따른 생존자 수는?**

In [67]:
df.query("survived == 1").groupby('deck').agg(survived = ('survived', 'count'))

survived
deck          
A            7
B           35
C           35
D           25
E           24
F            8
G            2

탑승지 이름 앞글자와 생존자 수에는 관계가 없다.

**7-1) 혼자 탑승한 승객과 가족과 함께 탑승한 승객 중 어느쪽의 생존율이 더 높을까?**

In [68]:
df.query("survived == 1 and alone == True")['survived'].count()
#혼자 탑승한 생존자 수

163

In [75]:
df.query("survived == 1 and alone == True")['survived'].count()/df.query("alone == True")['alone'].count()
#혼자 탑승한 생존자 비율 (혼자 탑승한 생존자 수/ 혼자 탑승한 승객 수)

0.30353817504655495

In [85]:
df.query("survived == 1 and alone == False")['survived'].count()
#가족과 함께 탑승한 생존자 수

179

In [84]:
df.query("survived == 1 and alone == False")['survived'].count()/df.query("alone == False")['alone'].count()
#가족과 함께 탑승한 생존자 비율

0.5056497175141242

혼자 탑승한 승객보다 가족과 함께 탑승한 승객의 생존율이 더 높다.

**7-2) 혼자 탑승한 사람 중 50세 이상인 사람의 생존자 수와 그 비율은?**

In [77]:
df.query("survived == 1 and alone == True and age >= 50")['survived'].count()
#혼자 탑승한 사람 중 50세 이상인 생존자 수

12

In [79]:
df.query("survived == 1 and alone == True and age >= 50")['survived'].count()/df.query("survived == 1 and age >= 50")['survived'].count()
#혼자 탑승한 사람 중 50세 이상인 생존자의 비율

0.4444444444444444

**8) 탑승지 이름의 앞글자에 따른 생존자 수는?**

In [80]:
df.query("survived == 1").groupby('embarked').agg(survived = ('survived', 'count'))

survived
embarked          
C               93
Q               30
S              217

탑승지 이름의 앞글자가 s인 탑승객의 생존자 수가 가장 많지만
탑승지 이름의 앞글자와 생존자 수의 관계는 없다.

**9) 형제 및 배우자와 함께 탑승한 승객의 생존율과 자녀 및 부모와 함께 탑승한 승객의 생존율 중 어느쪽이 더 높을까?**

In [90]:
df.query("survived == 1 and sibsp != 0")['survived'].count()/df.query("sibsp != 0")['survived'].count()
#형제 및 배우자와 함께 탑승한 승객의 생존율

0.4664310954063604

In [91]:
df.query("survived == 1 and parch != 0")['survived'].count()/df.query("parch != 0")['survived'].count()
#자녀 및 부모와 함께 탑승한 승객의 생존율

0.5117370892018779

자녀 및 부모와 함께 탑승한 승객의 생존율이 더 큰 것을 알 수 있다.

**10-1) 승객이 지불한 요금에 따른 생존율은?**

In [96]:
df.query("survived == 1 and fare <= 10")['survived'].count()/df.query("fare <= 10")['survived'].count()

0.19940476190476192

In [97]:
df.query("survived == 1 and fare > 10 and fare <= 30")['survived'].count()/df.query("fare > 10 and fare <= 30")['survived'].count()

0.43302180685358255

In [98]:
df.query("survived == 1 and fare > 30 and fare <= 70")['survived'].count()/df.query("fare > 30 and fare <= 70")['survived'].count() 

0.46511627906976744

In [100]:
df.query("survived == 1 and fare > 70")['survived'].count()/df.query("fare > 70")['survived'].count()  

0.7238095238095238

지불한 요금이 많은 승객의 무리일수록 생존율이 커지는 것을 보아
요금과 생존율은 대체적으로 비례한다고 볼 수 있다.

**10-2) 생존자가 지불한 요금의 평균은 전체 승객이 지불한 요금의 평균보다 큰가?**

In [101]:
df.query("survived == 1")['fare'].mean() 
#생존자가 지불한 요금의 평균

48.39540760233917

In [102]:
df['fare'].mean() 
#전체 승객이 지불한 요금의 평균

32.2042079685746

전체 승객이 지불한 요금의 평균보다 생존자가 지불한 요금의 평균이 더 큰 것을 알 수 있다.